🪞🚆 Symmetric LCWA

mberr opened the pull request:
Fix #1094

mberr committed:
add symmetric lcwa loop

mberr committed:
test + fixes

mberr committed:
reduce code duplication

mberr committed:
add comment

cthoyt commented on src/pykeen/training/lcwa.py:
```diff
@@ -237,3 +225,65 @@ def _check_slicing_availability(self, supports_sub_batching: bool):
+        slice_size: Optional[int] = None,
+    ) -> torch.FloatTensor:  # noqa: D102
+        # unpack
+        hrt_batch = batch[0]
```
Im confused by this line, why is only one thing unpacked?

mberr replied:
https://github.com/pykeen/pykeen/pull/1098#discussion_r958197972

cthoyt commented on src/pykeen/training/lcwa.py:
```diff
@@ -237,3 +225,65 @@ def _check_slicing_availability(self, supports_sub_batching: bool):
+# note: we use Tuple[Tensor] here, so we can re-use TensorDataset instead of having to create a custom one
+class SymmetricLCWATrainingLoop(TrainingLoop[Tuple[MappedTriples], Tuple[MappedTriples]]):
+    """
+    A "symmetric" LCWA scoring heads *and* tails at once.
```
Can we please get more high level documentation here? You already made a nice write up in the issue, just move it over here. Probably also link the issue itself

mberr replied:
https://github.com/pykeen/pykeen/pull/1098/commits/0e37df0e16aa5e98d270ec9365300fe855ca904e

mberr commented on src/pykeen/training/lcwa.py:
```diff
@@ -237,3 +225,65 @@ def _check_slicing_availability(self, supports_sub_batching: bool):
         raise MemoryError("The current model can't be trained on this hardware with these parameters.")
+
+
+# note: we use Tuple[Tensor] here, so we can re-use TensorDataset instead of having to create a custom one
```
here is the reason why I unpack the singleton tuple in `process_batch`

mberr committed:
copy issue description

cthoyt commented on src/pykeen/training/lcwa.py:
```diff
@@ -237,3 +225,77 @@ def _check_slicing_availability(self, supports_sub_batching: bool):
+    r"""
+    A "symmetric" LCWA scoring heads *and* tails at once.
+
+    This objective was introduced by [lacroix2018] as
```
don't forget the trailing underscore to actually link it

```suggestion
    This objective was introduced by [lacroix2018]_ as
```

mberr committed:
fix link

Co-authored-by: Charles Tapley Hoyt <cthoyt@gmail.com>

cthoyt commented on tests/test_training/test_training_loop.py:
```diff
@@ -85,3 +85,10 @@ class SoftPlusLCWATrainingLoopTestCase(cases.TrainingLoopTestCase):
+    """Test for symmetric LCWA with cross-entropy."""
+
+    cls = SymmetricLCWATrainingLoop
+    loss_cls = CrossEntropyLoss
```
is there a whitelist/blacklist of losses usable with Symmetric LCWA?

mberr replied:
All losses which support LCWA should also be able to support SymLCWA. And, afaik, all losses support LCWA.

cthoyt reviewed (approved):
Final thoughts before merging:

1. Are there concerns about what loss functions make sense for Symmetric LCWA? If so, let's document these
2. Could this "symmetric" trick also be applied to sLCWA
3. Are we going to start confusing sLCWA with Symmetric LCWA?

mberr commented on src/pykeen/training/lcwa.py:
I am not sure whether the name should stay like that, since the targets are different than in the normal LCWA loop.

mberr replied:
For now, I added a note that the name might change in the future: https://github.com/pykeen/pykeen/pull/1098/commits/91f6bc8f37300d6a96755745836806088805e026

mberr committed:
add note

mberr committed:
Merge branch 'master' into sym-lcwa

mberr committed:
update readme

mberr committed:
Fix SLCWATrainingLoop._create_training_data_loader

if shuffle was provided, it had been mistakenly forwarded to the outer
data loader
